### ISSUE TITLE
Collector calling reset before collecting

### ISSUE THREAD

**Issue by seermer** [aai-institute/tianshou#785]
- [x] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [x] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, gym, torch, numpy, sys
  print(tianshou.__version__, gym.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```
0.4.10 0.26.2 1.13.0+cu117 1.23.5 3.10.8 | packaged by conda-forge | (main, Nov 24 2022, 14:07:00) [MSC v.1916 64 bit (AMD64)] win32

Related issue: https://github.com/thu-ml/tianshou/issues/555#issuecomment-1059677918

This is not exactly a bug, nor exactly a feature request, but I think it is something that might be improved.

According to Gym official doc, reset should be called when "Resets the environment to an initial state and returns the initial observation."
In other words, end the previous episode (if any), and start a new one.
**Current behavior**
In the linked issue/comment above, two resets do not come with any interaction with the environment (so starting an environment episode is not intended).
This can introduce significant problems in some cases, especially when the environment itself is time-consuming or contains persistent internal data. For example, some video games may come with slight differences every time you start a new round based on previous episodes. Some environments cannot be easily restarted, and have to wait for a natural ending to start a new episode (and this may take a very long time or wastes many resources).

**Expected behavior**
End the current episode (if unfinished), and do not start a new one when unnecessary.

**Possible fix**
Since the gym library API does not come with anything that ends an episode without starting a new one, one possible fix would be to ask users to extend gym environments with an additional function (for example, let's call it end_episode()). 
If the given environment does have end_episode() implemented, call that instead of reset(), if not, call reset() just like before.

**Comment by Trinkle23897** [comment]
I think the easiest way is to add an env wrapper that has its own state, and if the last time `self.reset` is called, just bypass it.

**Comment by seermer** [comment]
Thanks for helping, I am still kind of confused, how do we determine if the current call to reset is the last time? 
Also, are there documentation for customizing the collection when working with tianshou's replay buffer? I might want to simply customize the environment interaction instead since I have little control to environment itself

**Comment by seermer** [comment]
After trying for a while, I personally believe this should be treated as a bug, or requires some special warning in the documentation.

So for example, let's say we want to collect a single episode with `collector.collect`. so here is what I'm doing:
```
env = ts.env.DummyVectorEnv(...)
model = Net(...)
optimizer = Adam(...)
policy = ts.policy.DQNPolicy(...)
buffer = ts.data.PrioritizedVectorReplayBuffer(...)
collector = ts.data.Collector(...)
collector.collect(n_episode=1, random=True)
```
What happens is `collector = ts.data.Collector(...)` will call reset() once, ignoring the output of reset(), now the environment starts, but the collection does not start (because `collector.collect` has not been called yet). Some environments will run once they started, they won't pause, most commonly seen in video games and some real-world environments.

Then, `collector.collect(n_episode=1, random=True)` is called, but it will not call  `reset() `before running the environment! Now the collector tries to step() before the environment even starts, or tries to collect an environment that has been left running without actions for a while (in other words, an environment that is not in its initial state).

After it tries to collect some possibly undefined/biased values returned by a possibly unstarted or halfway-started environment, it calls reset() at the end of `collector.collect()`. Now it starts a new environment without collecting anything from it.

So after running the above code, the environment ran more than once, but the buffer contains nothing but undefined/biased values.

**In conclusion,** when using environments that do not run step-by-step, the collection will almost always collect unusable (undefined/biased) data and waste plenty of extra resources.

**Comment by Trinkle23897** [comment]
```py
class IgnoreSecondReset(gym.Wrapper):
  def __init__(self, env):
    super.__init__(self, env)
    self.state = "step"
    self.last_reset_output = None
  def step(self, *args, **kwargs):
    self.state = "step"
    return self.env.step(*args, **kwargs)
  def reset(self, *args, **kwargs):
    if self.state == "reset":
      return self.last_reset_output
    self.state = "reset"
    self.last_reset_output = self.env.reset(*args, **kwargs)
    return self.last_reset_output

env = IgnoreSecondReset(gym.make(...))
# or env = DummyVectorEnv([lambda: IgnoreSecondReset(gym.make(...)) for _ in range(n)])

**Comment by seermer** [comment]
Thank you so much! I made some changes based on your code to ignore all reset() calls from the library, and manage environment reset on my own, it works as intended now.

Edit: actually, although it looked right, it seems like it still doesn't work as intended, since the wrapper reserves the output of the first reset and returns it the second time reset gets called. Correct me if I am wrong, but I believe in this case, the observation found at the beginning of the current episode will be stored as if it is found at the next episode (I read the source code, and saw that the return value of the reset at the end of an episode will be stored as self.data.next_obs).
I guess I will write my own data collectors and buffers instead, so I will leave this issue closed.
